relay-config/src/config.rs

```diff
@@ -882,41 +885,126 @@ pub enum TopicAssignment {
-        #[serde(rename = "name")]
-        topic_name: String,
-        /// An identifier referencing one of the kafka configurations in `secondary_kafka_configs`.
-        #[serde(rename = "config")]
-        kafka_config_name: String,
+    Secondary(KafkaTopicConfig),
+    /// If we want to configure multiple kafka clusters, we can create a mapping of the
+    /// range of logical shards to the kafka configuration.
+    Sharded(Sharded),
+}
+
+/// Configuration for topic
+#[derive(Serialize, Deserialize, Debug)]
+pub struct KafkaTopicConfig {
+    /// The topic name to use.
+    #[serde(rename = "name")]
+    topic_name: String,
+    /// The Kafka config name will be used to produce data to the given topic.
+    #[serde(rename = "config")]
+    kafka_config_name: String,
+}
+
+/// Configuration for logical shards -> kafka configuration mapping.
+/// The configuration for this should look like:
+///     ```
+///     metrics:
+///        shards: 65000
+///        mapping:
+///          0:
+///              name: "ingest-metrics-1"
+///              config: "metrics_1"
+///          25000:
+///              name: "ingest-metrics-2"
+///              config: "metrics_2"
+///          45000:
+///              name: "ingest-metrics-3"
+///              config: "metrics_3"
+///     ```
+///
+/// where the `shards` defines how many logical shards must be created, and `mapping`
+/// describes the per-shard configuration. Index in the `mapping` is the initial inclusive
+/// index of the shard and the range is last till the next index or the maximum shard defined in
+/// the `shards` option. The first index must always start with 0.
+#[derive(Serialize, Deserialize, Debug)]
+pub struct Sharded {
+    /// The number of shards used for this topic.
+    shards: u64,
+    /// The Kafka configuration assigned to the specific shard range.
+    mapping: BTreeMap<u64, KafkaTopicConfig>,
+}
+
+/// Contains kafka config names.
+#[derive(Debug)]
+enum KafkaConfigName<'a> {
+    /// The single kafka config name.
+    Single {
+        topic_name: &'a str,
+        kafka_config_name: Option<&'a str>,
+    },
+    /// Shard ranges mapped to the proper configs.
+    Sharded {
+        /// The maximum number of logical shards for this set of configs.
+        shards: u64,
+        /// The list of configs in relation to the shard ranges they will be used for.
+        configs: BTreeMap<u64, &'a KafkaTopicConfig>,
+    },
+}
+
+/// Describes Kafka config, with all the parameters extracted, which will be used for creating the
+/// kafka producer.
+#[derive(Debug)]
+pub enum KafkaConfig<'a> {
+    /// Single config with Kafka parameters.
+    Single(KafkaParams<'a>),
+
+    /// The list of the Kafka configs with related shard configs.
+    Sharded {
+        /// The maximum number of logical shards for this set of configs.
+        shards: u64,
+        /// The list of the sharded Kafka configs.
+        configs: BTreeMap<u64, KafkaParams<'a>>,
```

Comment:
We have this mapping in three different places now, I wonder if it would be possible to have it only once.

Reply:
@jjbayer I've tried to remove one of the levels of indirection to eliminate of of the repeating BTreeMap, see [0db194e](https://github.com/getsentry/relay/pull/1454/commits/0db194e5b2a53a2cd8462d8dc369c997a38f6439) 
Please, have another look if this is somewhat better approach.